EventFilter/Phase2TrackerRawToDigi/plugins/DTHDAQToFEDRawDataConverter.cc

```diff
@@ -0,0 +1,145 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "DataFormats/FEDRawData/interface/FEDRawData.h"
+#include "DataFormats/FEDRawData/interface/FEDRawDataCollection.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include <fstream>
+#include <vector>
+#include <iostream>
+#include <iomanip>
+#include <cstdint>
+
+// Constants for bit field widths, markers, and sizes
+constexpr unsigned int orbitHeaderSize = 32;
+constexpr unsigned int fragmentTrailerSize = 16;
+constexpr unsigned int fragmentPayloadWordSize = 16;  // Each fragment payload word is 16 bytes
+constexpr uint8_t orbitHeaderMarkerH = 0x48;
+constexpr uint8_t orbitHeaderMarkerO = 0x4F;
+constexpr uint8_t fragmentTrailerMarkerF = 0x48;
+constexpr uint8_t fragmentTrailerMarkerH = 0x46;
+
+class DTHDAQToFEDRawDataConverter : public edm::stream::EDProducer<> {
```

Comment:
If you are reading a txt file, it may not be thread safe, in which case using a stream::EDProducer which uses threads is dangerous. You may have to use a one::EDProducer, illustrated in https://github.com/cms-L1TK/cmssw/blob/L1TK-dev-14_2_0_pre4/L1Trigger/TrackFindingTracklet/plugins/L1FPGATrackProducer.cc#L141 .

Reply:
Done